extension/entitystore/serviceprovider.go

```diff
@@ -0,0 +1,263 @@
+// service name is retrieved based on the following priority chain
+//  1. Incoming telemetry attributes
+//  2. CWA config
+//  3. instance tags - The tags attached to the EC2 instance. Only scrape for tag with the following key: service, application, app
+//  4. IAM Role - The IAM role name retrieved through IMDS(Instance Metadata Service)
+func (s *serviceprovider) logFileServiceAttribute(logFile LogFileGlob, logGroup LogGroupName) ServiceAttribute {
+	return mergeServiceAttributes([]serviceAttributeProvider{
+		func() ServiceAttribute { return s.serviceAttributeForLogGroup(logGroup) },
+		func() ServiceAttribute { return s.serviceAttributeForLogFile(logFile) },
+		s.serviceAttributeFromImdsTags,
+		s.serviceAttributeFromIamRole,
+		s.serviceAttributeFromAsg,
+		s.serviceAttributeFallback,
+	})
+}
+
+func (s *serviceprovider) getServiceNameAndSource() (string, string) {
+	sa := mergeServiceAttributes([]serviceAttributeProvider{
+		s.serviceAttributeFromImdsTags,
+		s.serviceAttributeFromIamRole,
+		s.serviceAttributeFallback,
+	})
+	return sa.ServiceName, sa.ServiceNameSource
+}
+
+func (s *serviceprovider) serviceAttributeForLogGroup(logGroup LogGroupName) ServiceAttribute {
+	if logGroup == "" {
+		return ServiceAttribute{}
+	}
+
+	return s.logGroups[logGroup]
+}
+
+func (s *serviceprovider) serviceAttributeForLogFile(logFile LogFileGlob) ServiceAttribute {
+	if logFile == "" {
+		return ServiceAttribute{}
+	}
+
+	return s.logFiles[logFile]
+}
+
+func (s *serviceprovider) serviceAttributeFromImdsTags() ServiceAttribute {
+	if s.imdsServiceName == "" {
+		return ServiceAttribute{}
+	}
+
+	return ServiceAttribute{
+		ServiceName:       s.imdsServiceName,
+		ServiceNameSource: ServiceNameSourceResourceTags,
+	}
+}
+
+func (s *serviceprovider) serviceAttributeFromIamRole() ServiceAttribute {
+	if s.iamRole == "" {
+		return ServiceAttribute{}
+	}
+
+	return ServiceAttribute{
+		ServiceName:       s.iamRole,
+		ServiceNameSource: ServiceNameSourceClientIamRole,
+	}
+}
+
+func (s *serviceprovider) serviceAttributeFromAsg() ServiceAttribute {
+	if s.ec2Info == nil || s.ec2Info.AutoScalingGroup == "" {
+		return ServiceAttribute{}
+	}
+
+	return ServiceAttribute{
+		Environment: "ec2:" + s.ec2Info.AutoScalingGroup,
+	}
+}
+
+func (s *serviceprovider) serviceAttributeFallback() ServiceAttribute {
+	attr := ServiceAttribute{
+		ServiceName:       ServiceNameUnknown,
+		ServiceNameSource: ServiceNameSourceUnknown,
+	}
+	if s.mode == config.ModeEC2 {
+		attr.Environment = "ec2:default"
+	}
+
+	return attr
+}
+
+func (s *serviceprovider) getIAMRole() error {
+	iamRole, err := s.metadataProvider.InstanceProfileIAMRole()
+	if err != nil {
+		return err
+	}
+	iamRoleArn, err := arn.Parse(iamRole)
+	if err != nil {
+		return err
+	}
+	iamRoleResource := iamRoleArn.Resource
+	if strings.HasPrefix(iamRoleResource, INSTANCE_PROFILE) {
+		roleName := strings.TrimPrefix(iamRoleResource, INSTANCE_PROFILE)
+		s.iamRole = roleName
```

Comment:
This is on a separate goroutine from the `get` functions. Can we check to make sure there aren't any race conditions?

Reply:
Blocking comment